aspnetcore/signalr/streaming.md

```diff
@@ -0,0 +1,65 @@
+---
+title: Use streaming in ASP.NET Core SignalR
+author: rachelappel
+description: 
+manager: wpickett
+monikerRange: '>= aspnetcore-2.1'
+ms.author: rachelap
+ms.custom: mvc
+ms.date: 06/07/2018
+ms.prod: aspnet-core
+ms.technology: aspnet
+ms.topic: article
+uid: signalr/streaming
+---
+
+# Use streaming in ASP.NET Core SignalR
+
+By [Brennan Conroy](https://github.com/BrennanConroy)
+
+Streaming is useful for scenarios when you don't have all the data to send to the client yet, but will get it over time and want to update the client whenever new data is available.
```

Comment:
Let's start by explaining what "streaming" means here :).

> ASP.NET Core SignalR supports streaming return values of server methods. This is useful for scenarios where fragments of data will come in over time. When a return value is streamed to the client, each fragment is sent to the client as soon as it is available, rather than waiting for all the data to become available.

Also, I want to feel out this blurb to see how people think. It's a useful analogy if you're familiar with these C# concepts:

----

A streaming method is the SignalR equivalent of a C# iterator (using `yield` syntax). When calling methods locally, if the method returns an `IList<T>`, all the items must be fetched and put into the list before a return value can be produced. However, if the method returns `IEnumerable<T>` and uses the `yield` keyword in C#, items are made available to the caller right away. Consider the following example program:

```csharp
using System;
using System.Collections.Generic;
using System.Threading;

public class Program {
    public static IList<int> BlockingCounter()
    {
        var items = new List<int>();
        for (var i = 0; i < 3; i++)
        {
            Thread.Sleep(1000);
            items.Add(i);
        }
        return items;
    }

    public static IEnumerable<int> YieldingCounter()
    {
        for (var i = 0; i < 3; i++)
        {
            Thread.Sleep(1000);
            yield return i;
        }
    }

    public static void Main()
    {
        Console.WriteLine(DateTime.Now);
        foreach (var item in BlockingCounter())
        {
            Console.WriteLine($"Blocking: {item}");
            Console.WriteLine(DateTime.Now);
        }

        foreach (var item in YieldingCounter())
        {
            Console.WriteLine($"Yielding: {item}");
            Console.WriteLine(DateTime.Now);
        }
    }
}
```

When this program runs, the output will be something like this:

```
6/9/2018 8:57:47 AM
Blocking: 0
6/9/2018 8:57:50 AM
Blocking: 1
6/9/2018 8:57:50 AM
Blocking: 2
6/9/2018 8:57:50 AM
Yielding: 0
6/9/2018 8:57:51 AM
Yielding: 1
6/9/2018 8:57:52 AM
Yielding: 2
6/9/2018 8:57:53 AM
```

For the `Blocking` variant, there is a delay of 3 seconds before the first item is written, and then all three items are returned at once. For the `Yielding` variant, the first item is returned 1 second after calling the method and then each item comes a second after the previous item.

This is similar to how streaming works in SignalR. In SignalR, **even if you use `yield`**, the server will wait for all the items to be returned before sending anything to the client, but when you use the streaming pattern, you can return items as soon as they become available.